File: python/pytoon/decode/scanner.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List

from pytoon.constants import SPACE, TAB
from pytoon.types import BlankLineInfo, Depth, ParsedLine


class LineCursor:
    def __init__(self, lines: List[ParsedLine], blank_lines: List[BlankLineInfo] | None = None) -> None:
        self._lines = lines
        self._blank_lines = blank_lines or []
        self._index = 0

    def peek(self) -> ParsedLine | None:
        return self._lines[self._index] if self._index < len(self._lines) else None

    def next(self) -> ParsedLine | None:
        if self._index >= len(self._lines):
            return None
        line = self._lines[self._index]
        self._index += 1
        return line

    def current(self) -> ParsedLine | None:
        return self._lines[self._index - 1] if self._index > 0 else None

    def advance(self) -> None:
        self._index += 1

    def at_end(self) -> bool:
        return self._index >= len(self._lines)

    def peek_at_depth(self, depth: Depth) -> ParsedLine | None:
        line = self.peek()
        return line if line and line.depth == depth else None

    def get_blank_lines(self) -> List[BlankLineInfo]:
        return self._blank_lines

    @property
    def length(self) -> int:
        return len(self._lines)


def to_parsed_lines(source: str, indent_size: int, strict: bool) -> tuple[list[ParsedLine], list[BlankLineInfo]]:
    if not source.strip():
        return [], []

    parsed: list[ParsedLine] = []
    blanks: list[BlankLineInfo] = []

    for index, raw in enumerate(source.split('\n')):
        line_number = index + 1
        indent = 0
        while indent < len(raw) and raw[indent] == SPACE:
            indent += 1

        content = raw[indent:]

        if not content.strip():
            depth = indent // indent_size if indent_size else 0
            blanks.append(BlankLineInfo(line_number=line_number, indent=indent, depth=depth))
            continue

        if strict:
            leading_ws = ''
            for ch in raw:
                if ch in (SPACE, TAB):
                    leading_ws += ch
                else:
                    break
            if TAB in leading_ws:
                raise SyntaxError(f"Line {line_number}: Tabs are not allowed in indentation in strict mode")
            if indent_size and indent % indent_size != 0:
                raise SyntaxError(
                    f"Line {line_number}: Indentation must be exact multiple of {indent_size}, but found {indent} spaces",
                )

        depth = indent // indent_size if indent_size else 0
        parsed.append(ParsedLine(raw=raw, indent=indent, content=content, depth=depth, line_number=line_number))

    return parsed, blanks

File: packages/toon_julia/src/decode/scanner.jl
"""
LineCursor handles parsing through lines with tracking of position.
"""
mutable struct LineCursor
    lines::Vector{ParsedLine}
    index::Int
    blank_lines::Vector{BlankLineInfo}
    
    function LineCursor(lines::Vector{ParsedLine}, blank_lines::Vector{BlankLineInfo} = BlankLineInfo[])
        new(lines, 1, blank_lines)
    end
end

function get_blank_lines(cursor::LineCursor)::Vector{BlankLineInfo}
    return cursor.blank_lines
end

function peek(cursor::LineCursor)::Union{ParsedLine, Nothing}
    if cursor.index > length(cursor.lines)
        return nothing
    end
    return cursor.lines[cursor.index]
end

function next(cursor::LineCursor)::Union{ParsedLine, Nothing}
    if cursor.index > length(cursor.lines)
        return nothing
    end
    line = cursor.lines[cursor.index]
    cursor.index += 1
    return line
end

function current(cursor::LineCursor)::Union{ParsedLine, Nothing}
    if cursor.index > 1 && cursor.index <= length(cursor.lines) + 1
        return cursor.lines[cursor.index - 1]
    end
    return nothing
end

function advance(cursor::LineCursor)
    cursor.index += 1
end

function at_end(cursor::LineCursor)::Bool
    return cursor.index > length(cursor.lines)
end

function Base.length(cursor::LineCursor)::Int
    return length(cursor.lines)
end

function peek_at_depth(cursor::LineCursor, target_depth::Int)::Union{ParsedLine, Nothing}
    line = peek(cursor)
    return line !== nothing && line.depth == target_depth ? line : nothing
end

"""
Converts source string to parsed lines.
"""
struct ScanResult
    lines::Vector{ParsedLine}
    blank_lines::Vector{BlankLineInfo}
end

function to_parsed_lines(source::String, indent_size::Int, strict::Bool)::ScanResult
    if isempty(strip(source))
        return ScanResult(ParsedLine[], BlankLineInfo[])
    end
    
    lines = split(source, '\n')
    parsed = ParsedLine[]
    blank_lines = BlankLineInfo[]
    
    for (i, raw) in enumerate(lines)
        line_number = i
        indent = 0
        while indent < length(raw) && raw[indent + 1] == SPACE
            indent += 1
        end
        
        content = raw[(indent + 1):end]
        
        # Track blank lines
        if isempty(strip(content))
            depth = compute_depth_from_indent(indent, indent_size)
            push!(blank_lines, BlankLineInfo(line_number, indent, depth))
            continue
        end
        
        depth = compute_depth_from_indent(indent, indent_size)
        
        # Strict mode validation
        if strict
            # Check for tabs in leading whitespace
            whitespace_region = raw[1:min(indent, length(raw))]
            if occursin(TAB, whitespace_region)
                throw(ArgumentError("Line $(line_number): Tabs are not allowed in indentation in strict mode"))
            end
            
            # Check for exact multiples of indent_size
            if indent > 0 && indent % indent_size != 0
                throw(ArgumentError("Line $(line_number): Indentation must be exact multiple of $(indent_size), but found $(indent) spaces"))
            end
        end
        
        push!(parsed, ParsedLine(raw, depth, indent, content, line_number))
    end
    
    return ScanResult(parsed, blank_lines)
end

function compute_depth_from_indent(indent_spaces::Int, indent_size::Int)::Int
    return div(indent_spaces, indent_size)
end

